### PR TITLE
perf: Cache number of groups for overflow detection

### LINKED ISSUE #51: perf: Looping over maps is taking a lot of CPU time during merge
~30% CPU time spent on mapiterinit and mapiternext due to:
```
	for _, svc := range m.metrics.Services {
		for _, si := range svc.ServiceInstanceGroups {
			totalTransactionGroupsConstraint.add(len(si.TransactionGroups))
			totalServiceTransactionGroupsConstraint.add(len(si.ServiceTransactionGroups))
			totalSpanGroupsConstraint.add(len(si.SpanGroups))
		}
	}
```

- Cache the counts and update during merge instead.
- Add more tests to ensure no regressions are introduced as we fix the performance issue; since removing the above code didn't fail any tests surprisingly.

Potential gains (with the above code removed):
```
goos: linux
goarch: amd64
pkg: github.com/elastic/apm-aggregation/aggregators
cpu: 11th Gen Intel(R) Core(TM) i7-11800H @ 2.30GHz
                                       │ bench-out/before-cached-counts │ bench-out/before-cached-counts-remove │
                                       │             sec/op             │     sec/op      vs base               │
NDJSONSerial/go-2.0.0.ndjson-16                             29.65m ± 7%     16.72m ±  2%  -43.62% (p=0.002 n=6)
NDJSONSerial/nodejs-3.29.0.ndjson-16                       17.529m ± 1%     9.657m ±  7%  -44.91% (p=0.002 n=6)
NDJSONSerial/python-6.7.2.ndjson-16                         50.22m ± 2%     23.44m ±  2%  -53.33% (p=0.002 n=6)
NDJSONSerial/ruby-4.5.0.ndjson-16                           26.13m ± 2%     14.20m ±  2%  -45.68% (p=0.002 n=6)
NDJSONParallel/go-2.0.0.ndjson-16                           29.52m ± 1%     17.13m ±  2%  -41.99% (p=0.002 n=6)
NDJSONParallel/nodejs-3.29.0.ndjson-16                     17.286m ± 6%     9.574m ±  9%  -44.61% (p=0.002 n=6)
NDJSONParallel/python-6.7.2.ndjson-16                       50.50m ± 2%     23.35m ±  2%  -53.76% (p=0.002 n=6)
NDJSONParallel/ruby-4.5.0.ndjson-16                         26.25m ± 1%     14.92m ± 22%  -43.16% (p=0.002 n=6)
geomean                                                     28.71m          15.34m        -46.56%

                                       │ bench-out/before-cached-counts │ bench-out/before-cached-counts-remove │
                                       │              B/op              │      B/op        vs base              │
NDJSONSerial/go-2.0.0.ndjson-16                            11.49Mi ± 1%      11.54Mi ± 0%       ~ (p=0.065 n=6)
NDJSONSerial/nodejs-3.29.0.ndjson-16                       6.912Mi ± 0%      7.013Mi ± 2%  +1.45% (p=0.041 n=6)
NDJSONSerial/python-6.7.2.ndjson-16                        19.32Mi ± 1%      19.49Mi ± 1%  +0.92% (p=0.004 n=6)
NDJSONSerial/ruby-4.5.0.ndjson-16                          10.27Mi ± 1%      10.34Mi ± 0%  +0.74% (p=0.004 n=6)
NDJSONParallel/go-2.0.0.ndjson-16                          11.46Mi ± 1%      11.51Mi ± 1%       ~ (p=0.065 n=6)
NDJSONParallel/nodejs-3.29.0.ndjson-16                     6.915Mi ± 0%      6.970Mi ± 1%       ~ (p=0.065 n=6)
NDJSONParallel/python-6.7.2.ndjson-16                      19.41Mi ± 1%      19.49Mi ± 1%       ~ (p=0.240 n=6)
NDJSONParallel/ruby-4.5.0.ndjson-16                        10.26Mi ± 0%      10.33Mi ± 0%  +0.61% (p=0.002 n=6)
geomean                                                    11.21Mi           11.29Mi       +0.73%

                                       │ bench-out/before-cached-counts │ bench-out/before-cached-counts-remove │
                                       │           allocs/op            │    allocs/op     vs base              │
NDJSONSerial/go-2.0.0.ndjson-16                             139.4k ± 3%       141.8k ± 0%       ~ (p=0.065 n=6)
NDJSONSerial/nodejs-3.29.0.ndjson-16                        78.03k ± 0%       81.51k ± 3%  +4.46% (p=0.002 n=6)
NDJSONSerial/python-6.7.2.ndjson-16                         213.2k ± 2%       215.9k ± 1%       ~ (p=0.065 n=6)
NDJSONSerial/ruby-4.5.0.ndjson-16                           122.3k ± 0%       124.5k ± 0%  +1.81% (p=0.002 n=6)
NDJSONParallel/go-2.0.0.ndjson-16                           139.4k ± 2%       142.0k ± 0%  +1.85% (p=0.002 n=6)
NDJSONParallel/nodejs-3.29.0.ndjson-16                      78.03k ± 0%       80.29k ± 2%  +2.89% (p=0.002 n=6)
NDJSONParallel/python-6.7.2.ndjson-16                       215.6k ± 3%       217.3k ± 1%       ~ (p=0.589 n=6)
NDJSONParallel/ruby-4.5.0.ndjson-16                         122.3k ± 0%       124.6k ± 0%  +1.88% (p=0.002 n=6)
geomean                                                     130.0k            132.7k       +2.08%

```

### PR DESCRIPTION
**Pull request by carsonip**
- Fix performance issue due to looping over maps by caching the constraints in merger.
- Add a test for overflow due to global limit.

Benchstat:
```
goos: linux
goarch: amd64
pkg: github.com/elastic/apm-aggregation/aggregators
cpu: 11th Gen Intel(R) Core(TM) i7-11800H @ 2.30GHz
                                       │ bench-out/cache-counts-in-constraints-before │ bench-out/cache-counts-in-constraints-after │
                                       │                    sec/op                    │       sec/op         vs base                │
AggregateCombinedMetrics-16                                              2.071µ ±  4%          2.003µ ±  3%   -3.26% (p=0.007 n=10)
AggregateBatchSerial-16                                                  5.873µ ±  7%          5.179µ ±  5%  -11.82% (p=0.000 n=10)
AggregateBatchParallel-16                                                6.147µ ±  7%          5.655µ ±  6%   -8.01% (p=0.001 n=10)
CombinedMetricsEncoding-16                                               1.426µ ±  2%          1.427µ ±  4%        ~ (p=0.955 n=10)
CombinedMetricsToBatch-16                                                62.86µ ±  3%          62.21µ ±  2%        ~ (p=0.123 n=10)
EventToCombinedMetrics-16                                                401.2n ±  3%          380.8n ±  3%   -5.11% (p=0.001 n=10)
NDJSONSerial/go-2.0.0.ndjson-16                                          33.86m ±  9%          19.04m ±  1%  -43.76% (p=0.000 n=10)
NDJSONSerial/nodejs-3.29.0.ndjson-16                                     20.19m ± 21%          11.10m ± 38%  -45.01% (p=0.000 n=10)
NDJSONSerial/python-6.7.2.ndjson-16                                      59.54m ± 36%          26.75m ±  2%  -55.08% (p=0.000 n=10)
NDJSONSerial/ruby-4.5.0.ndjson-16                                        31.18m ± 31%          17.01m ±  8%  -45.45% (p=0.000 n=10)
NDJSONParallel/go-2.0.0.ndjson-16                                        34.87m ± 43%          20.21m ±  4%  -42.04% (p=0.000 n=10)
NDJSONParallel/nodejs-3.29.0.ndjson-16                                   22.85m ± 41%          11.34m ±  3%  -50.35% (p=0.000 n=10)
NDJSONParallel/python-6.7.2.ndjson-16                                    65.21m ± 34%          27.83m ±  3%  -57.33% (p=0.000 n=10)
NDJSONParallel/ruby-4.5.0.ndjson-16                                      32.89m ± 47%          16.73m ±  3%  -49.14% (p=0.000 n=10)
geomean                                                                  692.4µ                462.3µ        -33.24%

                                       │ bench-out/cache-counts-in-constraints-before │ bench-out/cache-counts-in-constraints-after │
                                       │                     B/op                     │        B/op         vs base                 │
AggregateCombinedMetrics-16                                            1.414Ki ± 1%          1.417Ki ±  0%       ~ (p=0.748 n=10)
AggregateBatchSerial-16                                                2.490Ki ± 2%          2.449Ki ±  2%  -1.63% (p=0.014 n=10)
AggregateBatchParallel-16                                              2.468Ki ± 2%          2.480Ki ±  3%       ~ (p=0.468 n=10)
CombinedMetricsEncoding-16                                               862.5 ± 7%            833.5 ± 42%       ~ (p=0.218 n=10)
CombinedMetricsToBatch-16                                              4.103Ki ± 0%          4.104Ki ±  0%       ~ (p=0.669 n=10)
EventToCombinedMetrics-16                                                0.000 ± 0%            0.000 ±  0%       ~ (p=1.000 n=10) ¹
NDJSONSerial/go-2.0.0.ndjson-16                                        11.55Mi ± 0%          11.62Mi ±  0%  +0.58% (p=0.003 n=10)
NDJSONSerial/nodejs-3.29.0.ndjson-16                                   6.958Mi ± 0%          6.995Mi ±  0%  +0.52% (p=0.000 n=10)
NDJSONSerial/python-6.7.2.ndjson-16                                    19.50Mi ± 1%          19.63Mi ±  0%  +0.65% (p=0.002 n=10)
NDJSONSerial/ruby-4.5.0.ndjson-16                                      10.34Mi ± 0%          10.41Mi ±  0%  +0.67% (p=0.000 n=10)
NDJSONParallel/go-2.0.0.ndjson-16                                      11.52Mi ± 0%          11.62Mi ±  0%  +0.84% (p=0.000 n=10)
NDJSONParallel/nodejs-3.29.0.ndjson-16                                 6.960Mi ± 0%          6.987Mi ±  0%  +0.38% (p=0.000 n=10)
NDJSONParallel/python-6.7.2.ndjson-16                                  19.52Mi ± 0%          19.58Mi ±  0%  +0.32% (p=0.015 n=10)
NDJSONParallel/ruby-4.5.0.ndjson-16                                    10.34Mi ± 0%          10.42Mi ±  0%  +0.72% (p=0.000 n=10)
geomean                                                                             ²                       +0.02%                ²
¹ all samples are equal
² summaries must be >0 to compute geomean

                                       │ bench-out/cache-counts-in-constraints-before │ bench-out/cache-counts-in-constraints-after │
                                       │                  allocs/op                   │     allocs/op       vs base                 │
AggregateCombinedMetrics-16                                              21.00 ± 0%             21.00 ± 0%       ~ (p=1.000 n=10) ¹
AggregateBatchSerial-16                                                  26.00 ± 4%             26.00 ± 4%       ~ (p=0.798 n=10)
AggregateBatchParallel-16                                                25.50 ± 6%             26.50 ± 6%       ~ (p=0.514 n=10)
CombinedMetricsEncoding-16                                               0.000 ± 0%             0.000 ± 0%       ~ (p=1.000 n=10) ¹
CombinedMetricsToBatch-16                                                95.00 ± 0%             95.00 ± 0%       ~ (p=1.000 n=10) ¹
EventToCombinedMetrics-16                                                0.000 ± 0%             0.000 ± 0%       ~ (p=1.000 n=10) ¹
NDJSONSerial/go-2.0.0.ndjson-16                                         139.4k ± 1%            142.0k ± 1%  +1.87% (p=0.000 n=10)
NDJSONSerial/nodejs-3.29.0.ndjson-16                                    78.00k ± 0%            79.11k ± 0%  +1.42% (p=0.000 n=10)
NDJSONSerial/python-6.7.2.ndjson-16                                     212.8k ± 1%            216.3k ± 1%  +1.64% (p=0.000 n=10)
NDJSONSerial/ruby-4.5.0.ndjson-16                                       122.2k ± 0%            124.4k ± 0%  +1.80% (p=0.000 n=10)
NDJSONParallel/go-2.0.0.ndjson-16                                       139.3k ± 0%            142.4k ± 1%  +2.18% (p=0.000 n=10)
NDJSONParallel/nodejs-3.29.0.ndjson-16                                  78.00k ± 0%            79.11k ± 0%  +1.42% (p=0.000 n=10)
NDJSONParallel/python-6.7.2.ndjson-16                                   214.1k ± 1%            214.3k ± 1%       ~ (p=0.912 n=10)
NDJSONParallel/ruby-4.5.0.ndjson-16                                     122.2k ± 0%            124.6k ± 0%  +1.92% (p=0.002 n=10)
geomean                                                                             ²                       +1.16%                ²
¹ all samples are equal
² summaries must be >0 to compute geomean

```

Closes #51

### CODE REVIEW

**Review by axw — APPROVED** [review]
whoa :)
LGTM!

File: aggregators/merger.go [hunk 1]
```diff
@@ -128,7 +117,7 @@ func (m *combinedMetricsMerger) merge(from *aggregationpb.CombinedMetrics) {
 func mergeServiceInstanceGroups(
 	to *serviceMetrics,
 	from []*aggregationpb.KeyedServiceInstanceMetrics,
-	totalTransactionGroupsConstraint, totalServiceTransactionGroupsConstraint, totalSpanGroupsConstraint *constraint,
+	totalTransactionGroupsConstraint, totalServiceTransactionGroupsConstraint, totalSpanGroupsConstraint *constraint.Constraint,
```
Comment (lahsivjar) [inline aggregators/merger.go:120]:
```suggestion
	constraints constraints,
```
Maybe pass the full constraints here?

**Review by lahsivjar — APPROVED** [review]
LGTM!
